### PR TITLE
chore: build noctilucent WASM library in a container

### PR DESCRIPTION
**Pull request by HBobertz**
Due to the addition of noctilucent to cdk, contributors needed to download rust/rustup to be able to build the cdk.

This uses the pre-existing dependency on Docker/Finch to containerize the process in order to not incur any further dependencies for contributors to manage.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -64,6 +64,7 @@ The following tools need to be installed on your system prior to installing the
 - [Python >= 3.6.5, < 4.0](https://www.python.org/downloads/release/python-365/)
 - [Docker >= 19.03](https://docs.docker.com/get-docker/)
   - the Docker daemon must also be running
+- [Rust >= 1.68, rustup >= 1.25.2](https://www.rust-lang.org/tools/install)
```
Comment (vinayak-kukreja) [inline CONTRIBUTING.md:67]: Thank you for adding this. 

How would users be informed of this change? I think an issue can be created and pinned informing the users about this, WDYT?

Comment (RomainMuller) [reply CONTRIBUTING.md:67]: Added a prerequisite test to inform users.

File: scripts/check-build-prerequisites.sh [hunk 2]
```diff
@@ -74,6 +74,19 @@ else
     wrong_version
 fi
 
+# Rustup
+app="rustup"
+app_min="1.25.2"
+check_which $app $app_min
+app_v=$(${app} --version 2>/dev/null)
+echo -e "Checking rustup version... \c"
+if [ $(echo $app_v | grep -c -E "^1\.(2(5\.[2-9]\d*|[6-9]\d*\.\d*)|[3-9]\d+\.\d+)") -eq 1 ]
```
Comment (Naumel) [inline scripts/check-build-prerequisites.sh:83]:
```suggestion
if [ $(echo $app_v | grep -c -E "^1\.(2(5\.[2-9]\d*|[6-9]\.\d*)|[3-9]\d+\.\d+)") -eq 1 ]
```

**Review by Naumel — DISMISSED** [review]
I think there's an extra `\d*`
~~we need to check > 2.5, meaning there is only one digit after the 6-9, before it flips from 2.* to 3.*.~~
the dot is before the 2, not after 👍

File: scripts/check-build-prerequisites.sh (hunk 2 above)
Comment (RomainMuller) [reply scripts/check-build-prerequisites.sh:83]: Here's a railroad diagram of the original (i.e: my) version of the expression (courtesy of https://regexper.com):

image

**Review by rix0rrr — CHANGES_REQUESTED** [review]
I don't think I'm cool with adding this dependency. Contributing is already hard, and this will make it even harder.

Why do we need it, exactly, and why can that not be addressed another way?

**Comment by RomainMuller** [comment]
Rust is already in `jsii/superchain`, and has been for a LONG while because some Python `cryptography` module that we transitively depend on needs to be built from it's rust source... I didn't think this would be much of a problem since it's already a host prerequisite in many cases...

I guess we can instead leverage the existing `docker` dependency instead and do this stuff within a container if you'd prefer that...

File: packages/aws-cdk/generate.sh [hunk 3]
```diff
@@ -15,30 +15,20 @@ cat > build-info.json <<HERE
...
-  cd ../../lib/vendor/noctilucent
-  rm package.json
-)
+# Build noctilucent package in a Docker/Finch VM
+NOCTILUCENT_GIT="https://github.com/iph/noctilucent.git"
+NOCTILUCENT_COMMIT_ID="6da7c9fade55f8443bba7b8fdfcd4ebfe5208fb1"
+if [ "$(cat lib/vendor/noctilucent/.version 2>/dev/null || echo '')" == "${NOCTILUCENT_GIT}:${NOCTILUCENT_COMMIT_ID}" ]
+then
+  echo "⏭️ Noctilucent WASM binary is up-to date, skipping build..."
+  echo "ℹ️ Delete lib/vendor/noctilucent/.version to force a rebuild."
```
Comment (rix0rrr) [inline packages/aws-cdk/generate.sh:24]: ❤️ clever

Comment (HBobertz) [reply packages/aws-cdk/generate.sh:24]: we were originally using git submodules but it turned out codepipeline didn't like that very much so we had to just get it ourselves

Comment (mrgrain) [reply packages/aws-cdk/generate.sh:24]: Submodules come with their own set of issues. If it's required for a build, we should not put it into a submodule.

The `noctilucent-wasm` npm package is still an option.

File: packages/aws-cdk/generate.sh [hunk 4]
```diff
@@ -15,30 +15,20 @@ cat > build-info.json <<HERE
...
-  wasm-pack build --target nodejs                                               \
-    --out-dir="${pkgroot}/lib/vendor/noctilucent"                               \
-    --out-name=index                                                            
-
-  cd ../../lib/vendor/noctilucent
-  rm package.json
-)
+# Build noctilucent package in a Docker/Finch VM
+NOCTILUCENT_GIT="https://github.com/iph/noctilucent.git"
+NOCTILUCENT_COMMIT_ID="6da7c9fade55f8443bba7b8fdfcd4ebfe5208fb1"
```
Comment (mrgrain) [inline packages/aws-cdk/generate.sh:20]: I agree with @rix0rrr this feels like we are reinventing versioning and releases.

**Review by rix0rrr — APPROVED** [review]

File: scripts/check-build-prerequisites.sh (hunk 2 above)
Comment (HBobertz) [reply scripts/check-build-prerequisites.sh]: cool site. Based on this it seems romains expression works. Can you confirm elena?

**Pull request merged**